File: backend/src/github/github.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Octokit } from '@octokit/rest';
import { RateLimiterService } from './rate-limiter.service';

export interface GitHubUser {
  login: string;
  id: number;
  name: string | null;
  email: string | null;
  bio: string | null;
  company: string | null;
  location: string | null;
  blog: string | null;
  avatar_url: string;
  public_repos: number;
  followers: number;
  following: number;
}

export interface GitHubRepo {
  name: string;
  full_name: string;
  description: string | null;
  language: string | null;
  stargazers_count: number;
  forks_count: number;
  html_url: string;
  pushed_at: string | null;
  created_at: string | null;
}

export interface GitHubOrgMember {
  login: string;
  id: number;
  avatar_url: string;
}

@Injectable()
export class GithubService {
  private readonly logger = new Logger(GithubService.name);
  private octokit: Octokit;

  constructor(
    private configService: ConfigService,
    private rateLimiter: RateLimiterService,
  ) {
    this.octokit = new Octokit({
      auth: this.configService.get('GITHUB_TOKEN'),
    });
  }

  private async waitForRateLimit(): Promise<void> {
    const check = await this.rateLimiter.canMakeRequest();

    if (!check.allowed && check.waitMs) {
      this.logger.warn(`Rate limited. Waiting ${check.waitMs}ms`);
      await new Promise((resolve) => setTimeout(resolve, check.waitMs));
    }
  }

  private extractRateLimitHeaders(
    headers: Record<string, unknown>,
  ): Record<string, string> {
    const result: Record<string, string> = {};

    if (headers['x-ratelimit-remaining'] !== undefined) {
      result['x-ratelimit-remaining'] = String(headers['x-ratelimit-remaining']);
    }
    if (headers['x-ratelimit-limit'] !== undefined) {
      result['x-ratelimit-limit'] = String(headers['x-ratelimit-limit']);
    }
    if (headers['x-ratelimit-reset'] !== undefined) {
      result['x-ratelimit-reset'] = String(headers['x-ratelimit-reset']);
    }

    return result;
  }

  async getUser(username: string): Promise<GitHubUser | null> {
    await this.waitForRateLimit();
    await this.rateLimiter.decrementRemaining();

    try {
      const response = await this.octokit.users.getByUsername({ username });
      await this.rateLimiter.updateFromHeaders(
        this.extractRateLimitHeaders(response.headers),
      );

      return response.data as GitHubUser;
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'status' in error) {
        const err = error as { status: number; response?: { headers: Record<string, unknown> } };
        if (err.status === 404) {
          return null;
        }
        if (err.response?.headers) {
          await this.rateLimiter.updateFromHeaders(
            this.extractRateLimitHeaders(err.response.headers),
          );
        }
      }
      this.logger.error(`Failed to fetch user ${username}:`, error);
      throw error;
    }
  }

  async getUserRepos(
    username: string,
    perPage = 30,
  ): Promise<GitHubRepo[]> {
    await this.waitForRateLimit();
    await this.rateLimiter.decrementRemaining();

    try {
      const response = await this.octokit.repos.listForUser({
        username,
        per_page: perPage,
        sort: 'updated',
        direction: 'desc',
      });

      await this.rateLimiter.updateFromHeaders(
        this.extractRateLimitHeaders(response.headers),
      );

      return response.data as GitHubRepo[];
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error) {
        const err = error as { response?: { headers: Record<string, unknown> } };
        if (err.response?.headers) {
          await this.rateLimiter.updateFromHeaders(
            this.extractRateLimitHeaders(err.response.headers),
          );
        }
      }
      this.logger.error(`Failed to fetch repos for ${username}:`, error);
      throw error;
    }
  }

  async getOrgMembers(
    org: string,
    perPage = 100,
  ): Promise<GitHubOrgMember[]> {
    await this.waitForRateLimit();
    await this.rateLimiter.decrementRemaining();

    try {
      const response = await this.octokit.orgs.listPublicMembers({
        org,
        per_page: perPage,
      });

      await this.rateLimiter.updateFromHeaders(
        this.extractRateLimitHeaders(response.headers),
      );

      return response.data as GitHubOrgMember[];
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error) {
        const err = error as { response?: { headers: Record<string, unknown> } };
        if (err.response?.headers) {
          await this.rateLimiter.updateFromHeaders(
            this.extractRateLimitHeaders(err.response.headers),
          );
        }
      }
      this.logger.error(`Failed to fetch members for org ${org}:`, error);
      throw error;
    }
  }

  async getAllOrgMembers(org: string): Promise<GitHubOrgMember[]> {
    const allMembers: GitHubOrgMember[] = [];
    let page = 1;
    const perPage = 100;

    while (true) {
      await this.waitForRateLimit();
      await this.rateLimiter.decrementRemaining();

      try {
        const response = await this.octokit.orgs.listPublicMembers({
          org,
          per_page: perPage,
          page,
        });

        await this.rateLimiter.updateFromHeaders(
          this.extractRateLimitHeaders(response.headers),
        );

        const members = response.data as GitHubOrgMember[];
        allMembers.push(...members);

        if (members.length < perPage) {
          break;
        }

        page++;
      } catch (error: unknown) {
        if (error && typeof error === 'object' && 'response' in error) {
          const err = error as { response?: { headers: Record<string, unknown> } };
          if (err.response?.headers) {
            await this.rateLimiter.updateFromHeaders(
              this.extractRateLimitHeaders(err.response.headers),
            );
          }
        }
        this.logger.error(`Failed to fetch members for org ${org} page ${page}:`, error);
        throw error;
      }
    }

    return allMembers;
  }

  async getRateLimitStatus(): Promise<{
    remaining: number;
    limit: number;
    resetAt: Date;
  }> {
    const response = await this.octokit.rateLimit.get();
    const core = response.data.rate;

    return {
      remaining: core.remaining,
      limit: core.limit,
      resetAt: new Date(core.reset * 1000),
    };
  }

  /**
   * Search for GitHub users by query
   * Used to find GitHub profiles from blog author names
   */
  async searchUsers(
    query: string,
    perPage = 5,
  ): Promise<GitHubOrgMember[]> {
    await this.waitForRateLimit();
    await this.rateLimiter.decrementRemaining();

    try {
      const response = await this.octokit.search.users({
        q: query,
        per_page: perPage,
      });

      await this.rateLimiter.updateFromHeaders(
        this.extractRateLimitHeaders(response.headers),
      );

      return response.data.items.map((user) => ({
        login: user.login,
        id: user.id,
        avatar_url: user.avatar_url,
      }));
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error) {
        const err = error as { response?: { headers: Record<string, unknown> } };
        if (err.response?.headers) {
          await this.rateLimiter.updateFromHeaders(
            this.extractRateLimitHeaders(err.response.headers),
          );
        }
      }
      this.logger.error(`Failed to search users with query "${query}":`, error);
      return [];
    }
  }
}

File: backend-java/src/main/java/com/peopleminer/github/GithubService.java
package com.peopleminer.github;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.kohsuke.github.*;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@Service
@Slf4j
public class GithubService {

    private final GitHub github;
    private final RateLimiterService rateLimiter;

    public GithubService(
            @Value("${github.token:}") String token,
            RateLimiterService rateLimiter
    ) throws IOException {
        this.rateLimiter = rateLimiter;
        if (token != null && !token.isEmpty()) {
            this.github = new GitHubBuilder().withOAuthToken(token).build();
            log.info("GitHub service initialized with authentication");
        } else {
            this.github = GitHub.connectAnonymously();
            log.warn("GitHub service initialized anonymously - rate limits will be very low");
        }
    }

    private void waitForRateLimit() throws InterruptedException {
        RateLimiterService.RateLimitCheck check = rateLimiter.canMakeRequest();
        if (!check.isAllowed() && check.getWaitMs() != null) {
            log.warn("Rate limited. Waiting {}ms", check.getWaitMs());
            Thread.sleep(check.getWaitMs());
        }
    }

    public GitHubUser getUser(String username) {
        try {
            waitForRateLimit();
            rateLimiter.decrementRemaining();

            GHUser user = github.getUser(username);
            return GitHubUser.builder()
                .login(user.getLogin())
                .id((int) user.getId())
                .name(user.getName())
                .email(user.getEmail())
                .bio(user.getBio())
                .company(user.getCompany())
                .location(user.getLocation())
                .blog(user.getBlog())
                .avatarUrl(user.getAvatarUrl())
                .publicRepos(user.getPublicRepoCount())
                .followers(user.getFollowersCount())
                .following(user.getFollowingCount())
                .build();
        } catch (GHFileNotFoundException e) {
            return null;
        } catch (Exception e) {
            log.error("Failed to fetch user {}: {}", username, e.getMessage());
            throw new RuntimeException("Failed to fetch user: " + username, e);
        }
    }

    public List<GitHubRepo> getUserRepos(String username, int limit) {
        try {
            waitForRateLimit();
            rateLimiter.decrementRemaining();

            GHUser user = github.getUser(username);
            List<GitHubRepo> repos = new ArrayList<>();

            for (GHRepository repo : user.listRepositories().withPageSize(limit)) {
                if (repos.size() >= limit) break;

                repos.add(GitHubRepo.builder()
                    .name(repo.getName())
                    .fullName(repo.getFullName())
                    .description(repo.getDescription())
                    .language(repo.getLanguage())
                    .stargazersCount(repo.getStargazersCount())
                    .forksCount(repo.getForksCount())
                    .htmlUrl(repo.getHtmlUrl().toString())
                    .pushedAt(toLocalDateTime(repo.getPushedAt()))
                    .createdAt(toLocalDateTime(repo.getCreatedAt()))
                    .build());
            }

            return repos;
        } catch (Exception e) {
            log.error("Failed to fetch repos for {}: {}", username, e.getMessage());
            throw new RuntimeException("Failed to fetch repos for: " + username, e);
        }
    }

    public List<GitHubOrgMember> getAllOrgMembers(String orgName) {
        try {
            List<GitHubOrgMember> allMembers = new ArrayList<>();
            GHOrganization org = github.getOrganization(orgName);

            for (GHUser member : org.listPublicMembers()) {
                waitForRateLimit();
                rateLimiter.decrementRemaining();

                allMembers.add(GitHubOrgMember.builder()
                    .login(member.getLogin())
                    .id((int) member.getId())
                    .avatarUrl(member.getAvatarUrl())
                    .build());
            }

            log.info("Found {} members in org {}", allMembers.size(), orgName);
            return allMembers;
        } catch (Exception e) {
            log.error("Failed to fetch members for org {}: {}", orgName, e.getMessage());
            throw new RuntimeException("Failed to fetch org members: " + orgName, e);
        }
    }

    public RateLimitStatus getRateLimitStatus() {
        try {
            GHRateLimit rateLimit = github.getRateLimit();
            GHRateLimit.Record core = rateLimit.getCore();

            return RateLimitStatus.builder()
                .remaining(core.getRemaining())
                .limit(core.getLimit())
                .resetAt(toLocalDateTime(core.getResetDate()))
                .build();
        } catch (Exception e) {
            log.error("Failed to get rate limit status: {}", e.getMessage());
            throw new RuntimeException("Failed to get rate limit status", e);
        }
    }

    private LocalDateTime toLocalDateTime(Date date) {
        if (date == null) return null;
        return date.toInstant().atZone(ZoneId.systemDefault()).toLocalDateTime();
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class GitHubUser {
        private String login;
        private int id;
        private String name;
        private String email;
        private String bio;
        private String company;
        private String location;
        private String blog;
        private String avatarUrl;
        private int publicRepos;
        private int followers;
        private int following;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class GitHubRepo {
        private String name;
        private String fullName;
        private String description;
        private String language;
        private int stargazersCount;
        private int forksCount;
        private String htmlUrl;
        private LocalDateTime pushedAt;
        private LocalDateTime createdAt;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class GitHubOrgMember {
        private String login;
        private int id;
        private String avatarUrl;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class RateLimitStatus {
        private int remaining;
        private int limit;
        private LocalDateTime resetAt;
    }
}
